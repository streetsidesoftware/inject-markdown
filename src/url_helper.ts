import fsPath from 'node:path';
import * as url from 'node:url';
import { posix } from 'path';

const isUrlRegExp = /^(https?|file):/i;

const urlHashRegExp = /#.*/;
const urlSearchRegExp = /[?][^#]*/;

export function isURL(url: unknown): url is URL {
    return url instanceof URL;
}

/**
 * Convert a path to a URL.
 * @param path - path
 * @param rel - Optional URL to be relative to.
 * @returns a URL
 */
export function pathToUrl(path: string | URL, rel?: URL): URL {
    if (isURL(path)) return path;
    if (isUrlRegExp.test(path)) return new URL(path, rel);

    const possibleUrl = new URL(path.replace(/\\/g, '/'), rel || url.pathToFileURL('./'));

    if (possibleUrl.protocol !== 'file:') {
        return possibleUrl;
    }

    // Remove params and hash.
    const pathOnly = path.replace(/[#?].*/g, '');
    const relDir = rel ? url.fileURLToPath(rel) : '';
    const resolved = fsPath.resolve(relDir, pathOnly);
    const pathUrl = url.pathToFileURL(resolved);
    if (!pathUrl.pathname.endsWith('/') && (pathOnly.endsWith('/') || pathOnly.endsWith('\\'))) {
        pathUrl.pathname += '/';
    }
    pathUrl.search = possibleUrl.search;
    pathUrl.hash = possibleUrl.hash;
    return pathUrl;
}

/**
 * Convert a directory path to a URL.
 * It ensures the resulting URL can be used as the root
 * of a relative URL.
 * @param dir - directory
 * @param rel - Optional URL to be relative to.
 * @returns a URL
 */
export function dirToUrl(dir: string | URL, rel?: URL): URL {
    const dirUrl = pathToUrl(dir, rel);
    if (!dirUrl.pathname.endsWith('/')) {
        const pathname = dirUrl.pathname + '/';
        if (dir === dirUrl) {
            return pathToUrl(pathname, dirUrl);
        }
        dirUrl.pathname = pathname;
    }
    return dirUrl;
}

/**
 * Determine the relative path.
 * @param fromUrl - from URL
 * @param toUrl - to URL
 * @returns the relative path
 */
export function relativePath(fromUrl: URL, toUrl: URL): RelURL {
    return new RelUrlImpl(posix.relative(fromUrl.pathname, toUrl.pathname), toUrl.search, toUrl.hash);
}

/**
 * The URL equivalent of `path.dirname`.
 * @param url - the url
 * @returns the parent URL.
 */
export function urlDirectory(url: URL): URL {
    return dirToUrl(posix.dirname(url.pathname), url);
}

export function parseRelativeUrl(url: string): RelURL {
    url = url.trim();
    const hash = url.match(urlHashRegExp)?.[0] || '';
    const search = url.match(urlSearchRegExp)?.[0] || '';
    const pathname = url.replace(/[?#].*/, '');
    return new RelUrlImpl(pathname, search, hash);
}

export interface RelURL {
    readonly pathname: string;
    readonly hash: string;
    readonly search: string;
    readonly href: string;
    toUrl(baseUrl: URL): URL;
}

class RelUrlImpl implements RelURL {
    private url: URL;

    constructor(readonly pathname: string, search: string, hash: string) {
        this.url = new URL('rel-url://');
        this.url.hash = hash;
        this.url.search = search;
    }

    get search() {
        return this.url.search;
    }

    get hash() {
        return this.url.hash;
    }

    get href() {
        const { pathname } = this;
        const { search, hash } = this.url;
        return pathname + search + hash;
    }

    toString(): string {
        return this.href;
    }

    toUrl(baseUrl: URL): URL {
        return new URL(this.href, baseUrl);
    }
}
