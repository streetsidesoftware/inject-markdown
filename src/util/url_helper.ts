import fsPath from 'node:path';
import * as url from 'node:url';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { posix } from 'path';

const isUrlRegExp = /^(https?|file):/i;

const urlHashRegExp = /#.*/;
const urlSearchRegExp = /[?][^#]*/;

export function isURL(url: unknown): url is URL {
    return url instanceof URL;
}

/**
 * Convert a path to a URL.
 * @param path - path
 * @param rel - Optional URL to be relative to.
 * @returns a URL
 */
export function pathToUrl(path: string | URL, rel?: URL): URL {
    if (isURL(path)) return path;
    if (isUrlRegExp.test(path)) return new URL(path, rel);

    const relUrl = parseRelativeUrl(path);
    return relUrl.toUrl(rel || url.pathToFileURL('./'));
}

/**
 * Convert a directory path to a URL.
 * It ensures the resulting URL can be used as the root
 * of a relative URL.
 * @param dir - directory
 * @param rel - Optional URL to be relative to.
 * @returns a URL
 */
export function dirToUrl(dir: string | URL, rel?: URL): URL {
    let dirUrl = pathToUrl(dir, rel);
    if (!dirUrl.pathname.endsWith('/')) {
        const pathname = dirUrl.pathname + '/';
        if (dir === dirUrl) {
            // Make a copy
            dirUrl = new URL(dirUrl.href);
        }
        dirUrl.pathname = pathname;
    }
    return dirUrl;
}

/**
 * Determine the relative path.
 * @param fromUrl - from URL
 * @param toUrl - to URL
 * @returns the relative path
 */
export function relativePath(fromUrl: URL, toUrl: URL): RelURL {
    return new RelUrlImpl(posix.relative(fromUrl.pathname, toUrl.pathname), toUrl.search, toUrl.hash);
}

/**
 * The URL equivalent of `path.dirname`.
 * @param url - the url
 * @returns the parent URL.
 */
export function urlDirectory(url: URL): URL {
    const clone = new URL(url.href);
    clone.pathname = posix.dirname(url.pathname);
    if (!clone.pathname.endsWith('/')) {
        clone.pathname += '/';
    }
    return clone;
}

export function parseRelativeUrl(url: string): RelURL {
    url = url.trim();
    const hash = url.match(urlHashRegExp)?.[0] || '';
    const search = url.match(urlSearchRegExp)?.[0] || '';
    const pathname = url.replace(/[?#].*/, '');
    return new RelUrlImpl(pathname, search, hash);
}

export interface RelURL {
    readonly pathname: string;
    readonly hash: string;
    readonly search: string;
    readonly href: string;
    toUrl(baseUrl: URL): URL;
}

class RelUrlImpl implements RelURL {
    private url: URL;

    constructor(readonly pathname: string, search: string, hash: string) {
        this.url = new URL('rel-url://');
        this.url.hash = hash;
        this.url.search = search;
    }

    get search() {
        return this.url.search;
    }

    get hash() {
        return this.url.hash;
    }

    get href() {
        const { pathname } = this;
        const { search, hash } = this.url;
        return pathname + search + hash;
    }

    toString(): string {
        return this.href;
    }

    toUrl(baseUrl: URL): URL {
        const href = this.href;
        if (isUrlRegExp.test(href)) return new URL(href, baseUrl);
        if (baseUrl.protocol === 'file:') {
            const baseFile = fileURLToPath(baseUrl);
            const baseDir = baseUrl.pathname.endsWith('/') ? baseFile : fsPath.dirname(baseFile);
            const filename = fsPath.resolve(baseDir, this.pathname);
            const url = pathToFileURL(filename);
            url.search = this.search;
            url.hash = this.hash;
            return url;
        }
        return new URL(this.href, baseUrl);
    }
}
