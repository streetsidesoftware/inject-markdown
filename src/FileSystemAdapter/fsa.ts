import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from '../util/url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        return await fetchUrl(file);
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}

async function fetchUrl(url: URL): Promise<string> {
    const response = await fetch(mapUrl(url));
    return await response.text();
}

function mapUrl(url: URL): string {
    let href = url.href;

    // Remap GitHub blobs
    href = href.replace(/^https?:\/\/github.com\/([^/]+\/[^/]+)\/blob\//, 'https://raw.githubusercontent.com/$1/');

    return href;
}
