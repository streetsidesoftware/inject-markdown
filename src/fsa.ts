import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from './url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        const response = await fetch(file.href);
        return await response.text();
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}
