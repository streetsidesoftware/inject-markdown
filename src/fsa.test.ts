import { describe, expect, test } from 'vitest';
import * as fsa from './fsa.js';
import { pathToUrl } from './url_helper.js';

describe('fsa', () => {
    test.each`
        file
        ${import.meta.url}
        ${'README.md'}
        ${'README.md#Amsterdam'}
        ${'https://raw.githubusercontent.com/streetsidesoftware/inject-markdown/df4cdb07d70f6d9247ea14a3ea4fa7b4512d329f/README.md'}
    `('readFile $file.toString()', async ({ file }) => {
        const fa = fsa.nodeFsa();
        const url = pathToUrl(file);
        await expect(fa.readFile(url, 'utf8')).resolves.toBeDefined();
    });

    test.each`
        file
        ${'https://raw.githubusercontent.com/streetsidesoftware/inject-markdown/df4cdb07d70f6d9247ea14a3ea4fa7b4512d329f/README.md#L5-L10'}
        ${'https://raw.githubusercontent.com/streetsidesoftware/inject-markdown/d7de2f5fe5f894df712c71d05eb3450ead944e73/src/app.mts#L22-L25'}
    `('readFile $file.toString()', async ({ file }) => {
        const fa = fsa.nodeFsa();
        const url = pathToUrl(file);
        await expect(fa.readFile(url, 'utf8')).resolves.toMatchSnapshot();
    });
});
