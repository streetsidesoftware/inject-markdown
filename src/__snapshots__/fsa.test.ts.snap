// Vitest Snapshot v1

exports[`fsa > readFile [Function toString]() 1`] = `
"# Inject Markdown

Inject files into a Markdown file.

## Justification

Sometimes it is necessary to assemble content into a static markdown file like \`README.md\`.
Manually copying and pasting content leads to duplication making it difficult to keep things in sync.

## Usage

Use HTML comments to mark where content will be injected.

\`\`\`markdown
<!--- @@inject: fixtures/sample-src.md --->
\`\`\`

\`\`\`sh
npx inject-markdown README.md
\`\`\`

## \`--help\`

\`\`\`sh
npx inject-markdown --help
\`\`\`

**Result**

<!--- @@inject: content/help.txt --->

\`\`\`
Usage: inject-markdown [options] <files...>

Inject file content into markdown files.

Arguments:
  files                 Files to scan for injected content.

Options:
  --no-must-find-files  No error if files are not found.
  --output-dir <dir>    Output Directory
  --cwd <dir>           Current Directory
  --clean               Remove the injected content.
  --verbose             Verbose output.
  --silent              Only output errors.
  --color               Force color.
  --no-color            Do not use color.
  -V, --version         output the version number
  -h, --help            display help for command
\`\`\`

<!--- @@inject-end: content/help.txt --->
"
`;

exports[`fsa > readFile [Function toString]() 2`] = `
"import { Command, CommanderError, program as defaultCommand } from 'commander';
import { globby, type Options as GlobbyOptions } from 'globby';
import { fileURLToPath } from 'node:url';
import * as path from 'path';
import { FileInjector, FileInjectorOptions } from './FileInjector.js';
import { nodeFsa } from './fsa.js';

const fs = nodeFsa();

const excludes = ['node_modules'];
const allowedFileExtensions: Record<string, boolean | undefined> = {
    '.md': true,
};

async function version(): Promise<string> {
    const pathSelf = fileURLToPath(import.meta.url);
    const pathPackageJson = path.join(path.dirname(pathSelf), '../package.json');
    const packageJson = JSON.parse(await fs.readFile(pathPackageJson, 'utf8'));
    return (typeof packageJson === 'object' && packageJson?.version) || '0.0.0';
}

interface Options extends FileInjectorOptions {
    mustFindFiles: boolean;
    cwd?: string;
}

async function findFiles(globs: string[], cwd: string | undefined) {
    const _cwd = process.cwd();
    cwd && process.chdir(cwd);
    const options: GlobbyOptions = {
        ignore: excludes,
        onlyFiles: true,
    };
    const files = await globby(
        globs.map((a) => a.trim()).filter((a) => !!a),
        options
    );
    process.chdir(_cwd);
    // console.log('%o', files);
    return files.filter((f) => path.extname(f) in allowedFileExtensions);
}

async function processGlobs(globs: string[], options: Options): Promise<boolean> {
    if (!globs.length) return false;

    const files = await findFiles(globs, options.cwd);

    if (!files.length) {
        return !options.mustFindFiles;
    }

    const injector = new FileInjector(fs, options);

    for (const file of files) {
        const r = await injector.processFile(file);
    }

    return true;
}

async function app(program = defaultCommand, argv?: string[]) {
    program
        .name('inject-markdown')
        .description('Inject file content into markdown files.')
        .argument('<files...>', 'Files to scan for injected content.')
        .option('--no-must-find-files', 'No error if files are not found.')
        .option('--output-dir <dir>', 'Output Directory')
        .option('--cwd <dir>', 'Current Directory')
        .option('--clean', 'Remove the injected content.')
        .option('--verbose', 'Verbose output.')
        .option('--silent', 'Only output errors.')
        .option('--color', 'Force color.')
        .option('--no-color', 'Do not use color.')
        .version(await version())
        .action(async (files: string[], options: Options, _command: Command) => {
            // console.log('Options: %o', options);
            const result = await processGlobs(files, options);
            if (!result) {
                throw new CommanderError(1, 'Not Found', 'No files found.');
            }
        });

    program.showHelpAfterError();

    return program.parseAsync(argv);
}

export async function run(program?: Command): Promise<void> {
    try {
        await app(program);
    } catch (e) {
        process.exitCode = 1;
        if (!(e instanceof CommanderError)) {
            console.log(e);
            return;
        }
        console.log(e.message);
    }
}
"
`;
